google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb

```diff
@@ -0,0 +1,375 @@
+
+              @push_thread_pool.shutdown
+              @push_thread_pool.wait_for_termination
+            end
+
+            self
+          end
+
+          ##
+          # @private
+          def acknowledge *messages
+            ack_ids = coerce_ack_ids messages
+            return true if ack_ids.empty?
+
+            async_acknowledger # To make sure the object is loaded
+
+            synchronize do
+              @async_acknowledger.acknowledge ack_ids
+              @inventory.remove ack_ids
+              unpause_streaming!
+            end
+
+            true
+          end
+
+          ##
+          # @private
+          def delay deadline, *messages
+            mod_ack_ids = coerce_ack_ids messages
+            return true if mod_ack_ids.empty?
+
+            async_delayer # To make sure the object is loaded
+
+            synchronize do
+              @async_delayer.delay deadline, mod_ack_ids
+              @inventory.remove mod_ack_ids
+              unpause_streaming!
+            end
+
+            true
+          end
+
+          def async_acknowledger
+            synchronize { @async_acknowledger ||= AsyncAcknowledger.new self }
+          end
+
+          def async_delayer
+            synchronize { @async_delayer ||= AsyncDelayer.new self }
+          end
+
+          def push request
+            synchronize { @request_queue.push request }
+          end
+
+          def inventory
+            synchronize { @inventory }
+          end
+
+          ##
+          # @private
+          def delay_inventory!
+            async_delayer # To make sure the object is loaded
+
+            synchronize do
+              return true if @inventory.empty?
+
+              @async_delayer.delay subscriber.deadline, @inventory.ack_ids
+            end
+
+            true
+          end
+
+          # @private
+          def to_s
+            format "(inventory: %i)", inventory.count
+          end
+
+          # @private
+          def inspect
+            "#<#{self.class.name} #{self}>"
+          end
+
+          protected
+
+          # rubocop:disable all
+
+          def background_run enum
+            until synchronize { @stopped }
+              synchronize do
+                if @paused
+                  @pause_cond.wait
+                  next
+                end
+              end
+
+              begin
+                # Cannot syncronize the enumerator, causes deadlock
+                response = enum.next
+                response.received_messages.each do |rec_msg_grpc|
+                  rec_msg = ReceivedMessage.from_grpc(rec_msg_grpc, self)
+                  synchronize do
+                    @inventory.add rec_msg.ack_id
+
+                    perform_callback_async rec_msg
+                  end
+                end
+                synchronize { pause_streaming! }
+              rescue StopIteration
+                break
+              end
+            end
+          rescue GRPC::DeadlineExceeded
```

Comment:
This should be the same logic as for Unavailable and Cancelled errors. I would have them all do exponential backoff with a min of immediately. If for some reason you're receiving deadline exceeded because of a server error, which is possible, you would not want to retry as fast as possible.